### PR TITLE
[SPARK-40197][SQL] Replace query plan with context for MULTI_VALUE_SUBQUERY_ERROR

### PR DESCRIPTION
**Pull request by vitaliili-db**
### What changes were proposed in this pull request?

Replace query plan with context for `MULTI_VALUE_SUBQUERY_ERROR`

### Why are the changes needed?

User-friendly errors.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by vitaliili-db** [comment]
@gengliangwang @MaxGekk

File: core/src/main/scala/org/apache/spark/SparkException.scala [hunk 1]
```diff
@@ -29,7 +29,8 @@ class SparkException(
     cause: Throwable,
     errorClass: Option[String],
     errorSubClass: Option[String],
-    messageParameters: Array[String])
+    messageParameters: Array[String],
+    context: Array[QueryContext] = Array.empty)
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/SparkException.scala:33]: @MaxGekk As you mentioned, you plan to add context in `SparkException`. Is this part looking good to you?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -2071,9 +2071,14 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
...
 
-  def multipleRowSubqueryError(plan: String): Throwable = {
+  def multipleRowSubqueryError(context: SQLQueryContext): Throwable = {
     new SparkException(
-      errorClass = "MULTI_VALUE_SUBQUERY_ERROR", messageParameters = Array(plan), cause = null)
+      errorClass = "MULTI_VALUE_SUBQUERY_ERROR",
+      messageParameters = Array.empty,
+      cause = null,
+      context = getQueryContext(context),
+      summary = getSummary(context)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2080]: Why do we need to get the summary here?

Comment (vitaliili-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2080]: `getSummary` is a method of `QueryErrorsBase`. `SparkException` does not inherit from it.

File: core/src/main/scala/org/apache/spark/SparkException.scala (hunk 1 above)
Comment (MaxGekk) [reply core/src/main/scala/org/apache/spark/SparkException.scala:33]: Yep.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -2071,9 +2071,14 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
...
+    )
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2081]: minor but in any case, ) should be on the line above, see https://github.com/databricks/scala-style-guide. For example:
```scala
new Bar(
  someVeryLongFieldName,  // 2 space indent here
  andAnotherVeryLongFieldName,
  "this is a string",
  3.1415)
```

**Review by MaxGekk — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/SparkException.scala [hunk 4]
```diff
@@ -50,6 +51,21 @@ class SparkException(
...
+      context: Array[QueryContext],
+      summary: String) =
+    this(
+      message = SparkThrowableHelper.getMessage(errorClass, null, messageParameters, summary),
+      cause = cause,
+      errorClass = Some(errorClass),
+      errorSubClass = None,
+      messageParameters = messageParameters,
+      context
+    )
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/SparkException.scala:67]: nit: see the code around
```suggestion
      context)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 3 above)
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2081]:
```suggestion
      summary = getSummary(context))
```

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 5]
```diff
@@ -606,30 +606,23 @@ class QueryExecutionErrorsSuite
     }
   }
 
-  test("MULTI_VALUE_SUBQUERY_ERROR: " +
-    "more than one row returned by a subquery used as an expression") {
+  test(
+    "MULTI_VALUE_SUBQUERY_ERROR: " +
+    "More than one row returned by a subquery used as an expression"
+  ) {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:612]:
```suggestion
    "More than one row returned by a subquery used as an expression") {
```

**Review by gengliangwang — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @vitaliili-db and @gengliangwang for review.

**Pull request closed without merging**